File: DSA/Stack/Code/ValidParantheses.java
// Valid Parentheses in Java

import java.util.Stack;

public class ValidParentheses {
    static boolean isValidParentheses(String str) {
        Stack<Character> stack = new Stack<>();
        
        for (char c : str.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (stack.isEmpty()) return false;
                
                char top = stack.pop();
                
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
    
    public static void main(String[] args) {
        System.out.println("=== Valid Parentheses ===\n");
        
        String[] tests = {
            "()",
            "(){}[]",
            "([{}])",
            "([)]",
            "{[}",
            "(((",
            ")"
        };
        
        for (String test : tests) {
            System.out.println("Expression: " + test + " -> " + 
                             (isValidParentheses(test) ? "Valid" : "Invalid"));
        }
        
        System.out.println("\nComplexity: O(n)");
        System.out.println("Space: O(n)");
    }
}

File: DSA/Stack/Code/ValidParantheses.cs
// Valid Parentheses in C#

using System;
using System.Collections.Generic;

class ValidParentheses {
    static bool IsValidParentheses(string str) {
        Stack<char> stack = new Stack<char>();
        
        foreach (char c in str) {
            if (c == '(' || c == '{' || c == '[') {
                stack.Push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (stack.Count == 0) return false;
                
                char top = stack.Pop();
                
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }
        
        return stack.Count == 0;
    }
    
    static void Main() {
        Console.WriteLine("=== Valid Parentheses ===\n");
        
        string[] tests = {
            "()",
            "(){}[]",
            "([{}])",
            "([)]",
            "{[}",
            "(((",
            ")"
        };
        
        foreach (string test in tests) {
            Console.WriteLine("Expression: " + test + " -> " + 
                            (IsValidParentheses(test) ? "Valid" : "Invalid"));
        }
        
        Console.WriteLine("\nComplexity: O(n)");
        Console.WriteLine("Space: O(n)");
    }
}
